bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/PBEKeySpecCommon.kt

```diff
@@ -0,0 +1,179 @@
+     * the new `PBEKeySpec` object.
+     *
+     * @param password the password.
+     */
+    constructor(password: CharArray?) {
+        if (password == null || password.size == 0) {
+            this.password = CharArray(0)
+        } else {
+            this.password = password.copyOf()
+        }
+    }
+
+    /**
+     * Constructor that takes a password, salt, iteration count, and
+     * to-be-derived key length for generating PBEKey of variable-key-size
+     * PBE ciphers.  An empty char[] is used if null is specified for
+     * `password`.
+     *
+     *
+     *  Note: the `password` and `salt`
+     * are cloned before they are stored in
+     * the new `PBEKeySpec` object.
+     *
+     * @param password the password.
+     * @param salt the salt.
+     * @param iterationCount the iteration count.
+     * @param keyLength the to-be-derived key length.
+     * @exception NullPointerException if `salt` is null.
+     * @exception IllegalArgumentException if `salt` is empty,
+     * i.e. 0-length, `iterationCount` or
+     * `keyLength` is not positive.
+     */
+    actual constructor(
+        password: CharArray?,
+        salt: ByteArray?,
+        iterationCount: Int,
+        keyLength: Int
+    ) {
+        if (password == null || password.size == 0) {
+            this.password = CharArray(0)
+        } else {
+            this.password = password.copyOf()
+        }
+        if (salt == null) {
+            throw NullPointerException(
+                "the salt parameter " +
+                    "must be non-null"
+            )
+        } else {
+            require(salt.size != 0) {
+                "the salt parameter " +
+                    "must not be empty"
+            }
+        }
+        this.salt = salt.copyOf()
+        require(iterationCount > 0) { "invalid iterationCount value" }
+        require(keyLength > 0) { "invalid keyLength value" }
+        this.iterationCount = iterationCount
+        this.keyLength = keyLength
+    }
+
+    /**
+     * Constructor that takes a password, salt, iteration count for
+     * generating PBEKey of fixed-key-size PBE ciphers. An empty
+     * char[] is used if null is specified for `password`.
+     *
+     *
+     *  Note: the `password` and `salt`
+     * are cloned before they are stored in the new
+     * `PBEKeySpec` object.
+     *
+     * @param password the password.
+     * @param salt the salt.
+     * @param iterationCount the iteration count.
+     * @exception NullPointerException if `salt` is null.
+     * @exception IllegalArgumentException if `salt` is empty,
+     * i.e. 0-length, or `iterationCount` is not positive.
+     */
+    constructor(password: CharArray?, salt: ByteArray?, iterationCount: Int) {
+        if (password == null || password.size == 0) {
+            this.password = CharArray(0)
+        } else {
+            this.password = password.copyOf()
+        }
+        if (salt == null) {
```

Comment:
Since we're re-implementing this, does it make sense to drop the nullable for `salt` so we don't need the null check here?

Reply:
yeah, good point